internal/server/singleprocess/state/task.go

```diff
@@ -0,0 +1,317 @@
+		taskId = r.Id
+	case *pb.Ref_Task_JobId:
+		s.log.Info("looking up task by job id", "job_id", r.JobId)
+		// Look up Task by jobid
+		task, err := s.taskByJobId(r.JobId)
+		if err != nil {
+			return nil, err
+		}
+
+		s.log.Info("found task id", "id", task.Id)
+		taskId = task.Id
+	default:
+		return nil, status.Error(codes.FailedPrecondition, "No valid ref id provided in Task ref to taskGet")
+	}
+
+	return &result, dbGet(b, []byte(strings.ToLower(taskId)), &result)
+}
+
+func (s *State) taskList(
+	memTxn *memdb.Txn,
+) ([]*pb.Ref_Task, error) {
+	iter, err := memTxn.Get(taskIndexTableName, taskIndexId+"_prefix", "")
+	if err != nil {
+		return nil, err
+	}
+
+	var result []*pb.Ref_Task
+	for {
+		next := iter.Next()
+		if next == nil {
+			break
+		}
+		idx := next.(*taskIndexRecord)
+
+		result = append(result, &pb.Ref_Task{
+			Ref: &pb.Ref_Task_Id{
+				Id: idx.Id,
+			},
+		})
+	}
+
+	return result, nil
+}
+
+func (s *State) taskDelete(
+	dbTxn *bolt.Tx,
+	memTxn *memdb.Txn,
+	ref *pb.Ref_Task,
+) error {
+	// Get the task. If it doesn't exist then we're successful.
+	_, err := s.taskGet(dbTxn, memTxn, ref)
+	if err != nil {
+		if status.Code(err) == codes.NotFound {
+			return nil
+		}
+
+		return err
+	}
+
+	// Delete from bolt
+	id, err := s.taskIdByRef(ref)
+	if err != nil {
+		return err
+	}
+
+	if err := dbTxn.Bucket(taskBucket).Delete(id); err != nil {
+		return err
+	}
+
+	// Delete from memdb
+	if err := memTxn.Delete(taskIndexTableName, &taskIndexRecord{Id: string(id)}); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// taskIndexSet writes an index record for a single task.
+func (s *State) taskIndexSet(txn *memdb.Txn, id []byte, value *pb.Task) error {
+	record := &taskIndexRecord{
+		Id: string(id),
+	}
+
+	// Insert the index
+	return txn.Insert(taskIndexTableName, record)
+}
+
+// taskIndexInit initializes the task index from persisted data.
+func (s *State) taskIndexInit(dbTxn *bolt.Tx, memTxn *memdb.Txn) error {
+	bucket := dbTxn.Bucket(taskBucket)
+	return bucket.ForEach(func(k, v []byte) error {
+		var value pb.Task
+		if err := proto.Unmarshal(v, &value); err != nil {
+			return err
+		}
+
+		if err := s.taskIndexSet(memTxn, k, &value); err != nil {
+			return err
+		}
+
+		return nil
+	})
+}
+
+func (s *State) taskId(t *pb.Task) []byte {
+	return []byte(strings.ToLower(t.Id))
+}
+
+func (s *State) taskIdByRef(ref *pb.Ref_Task) ([]byte, error) {
+	var taskId string
+	switch t := ref.Ref.(type) {
+	case *pb.Ref_Task_Id:
+		taskId = t.Id
+	case *pb.Ref_Task_JobId:
+		// Look up Task by jobid
+		task, err := s.taskByJobId(t.JobId)
+		if err != nil {
+			return nil, err
+		}
+
+		taskId = task.Id
+	default:
+		return nil, status.Error(codes.FailedPrecondition, "No valid ref id provided in Task ref to taskIdByRef")
+	}
+
+	return []byte(strings.ToLower(taskId)), nil
+}
+
+func (s *State) taskByJobId(jobId string) (*pb.Task, error) {
+	trackedTasks, err := s.TaskList()
```

Comment:
Ah hah, I would add a `jobId` field to the index and use an index lookup for this. I think we'll regularly have a lot of tasks and this is going to get nasty.

Reply:
Ah yeah, good call!